lib/Dialect/TritonGPU/Transforms/OptimizeThreadLocality.cpp

```diff
@@ -1,20 +1,95 @@
 #include "mlir/Pass/Pass.h"
 #include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 #include "mlir/Transforms/Passes.h"
 #include "triton/Dialect/TritonGPU/IR/Dialect.h"
 #include "triton/Dialect/TritonGPU/Transforms/Passes.h"
 #include "triton/Dialect/TritonGPU/Transforms/Utility.h"
 #include <memory>
+#include <numeric>
 #define GEN_PASS_CLASSES
 #include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
 
 using namespace mlir;
+namespace {
+// Change the layout of view destination when used by a reduction in order to
+// minimize the amount of cross thread communication for the reduction.
+struct OptimizeViewLayoutPattern
+    : public mlir::OpRewritePattern<triton::ReshapeOp> {
+  OptimizeViewLayoutPattern(mlir::MLIRContext *context)
+      : OpRewritePattern<triton::ReshapeOp>(context, 1) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(triton::ReshapeOp viewOp,
+                  mlir::PatternRewriter &rewriter) const override {
+    if (!viewOp.getAllowReorder())
+      return failure();
+    std::optional<int> reductionAxis;
+    for (Operation *user : viewOp.getResult().getUsers()) {
+      if (auto reduceOp = dyn_cast<triton::ReduceOp>(user)) {
+        if (reductionAxis) {
+          if (reductionAxis != reduceOp.getAxis())
+            return failure();
+        } else {
+          reductionAxis = reduceOp.getAxis();
+        }
+      }
+    }
+    if (!reductionAxis)
+      return failure();
+    auto tensorType = viewOp.getResult().getType().cast<RankedTensorType>();
+    ArrayRef<int64_t> shape = tensorType.getShape();
+    llvm::SmallVector<unsigned> order;
+    for (int i = shape.size() - 1; i >= 0; i--) {
+      if (i != *reductionAxis)
+        order.push_back(i);
+    }
+    // Make the reduction axis last so that elements won't be distributed
+    // amongst threads along this dimension.
+    order.push_back(*reductionAxis);
+    llvm::SmallVector<unsigned> sizePerThread(shape.size(), 1);
```

Comment:
Suppose the tensor has 1024 total elements and we have 128 threads per block.  So each thread is going to get 1024 / 128 = 8 elements.

Suppose the shape of the tensor is `[128, 8]` and we're reducing along the size-8 dimension.  In this case, if we get to choose the layout, I'd think we would want to choose sizePerThread = `[1,8]` so we can do the reduction entirely within one thread.

Reply:
so in this case if we have 2 warps this would generate such a layout:
```
 #triton_gpu.blocked<{sizePerThread = [1, 1], threadsPerWarp = [32, 1], warpsPerCTA = [2, 1], order = [0, 1], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [0, 1]}>
```

so threadsPerWarp[1] = 1 and warpsPerCTA[1] = 1 which means all the elements of dim 1 are still going to be on the same thread, it is just the order of elements in the thread that is going to be different (which doesn't matter).

I had started with the solution you described but I felt like I was more likely to generate expensive views that way so I went with this. I think it should always make all the elements being reduced in one thread unless it is a case where the layout has more elements than the tensor.